### ISSUE TITLE
DecoderException : java.lang.IndexOutOfBoundsException @ io.netty.handler.codec.MessageToMessageDecoder:99 (with 1.9 client)

### ISSUE THREAD

**Issue by L4BORG** [MylesIsCool/ViaVersion#40]
This is still an issue with 1.9 client, 1.8 works fine now... 1.9 client would only see: 

```
DecoderException : java.lang.IndexOutOfBoundsException @ io.netty.handler.codec.MessageToMessageDecoder:99
```

and with debug=true I'd get:
```
[18:37:16] [Thread-22/INFO]: UUID of player player is 2b90d869-0140-3bee-a66f-75a0d0941663
[18:37:16] [Server thread/INFO]: player[/127.0.0.1:38097] logged in with entity id 27 at ([world]-1.6557193171111857, 64.0, -0.08174551224638282)
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.io.IOException: Error while write(...): Broken pipe
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.epoll.Native.writeAddress(Native Method)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.epoll.EpollSocketChannel.writeBytes(EpollSocketChannel.java:125)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.epoll.EpollSocketChannel.doWriteSingle(EpollSocketChannel.java:337)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.epoll.EpollSocketChannel.doWrite(EpollSocketChannel.java:325)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.AbstractChannel$AbstractUnsafe.flush0(AbstractChannel.java:707)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.flush0(AbstractEpollChannel.java:240)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.AbstractChannel$AbstractUnsafe.flush(AbstractChannel.java:676)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.DefaultChannelPipeline$HeadContext.flush(DefaultChannelPipeline.java:1059)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:688)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:669)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.ChannelOutboundHandlerAdapter.flush(ChannelOutboundHandlerAdapter.java:115)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:688)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:669)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.ChannelOutboundHandlerAdapter.flush(ChannelOutboundHandlerAdapter.java:115)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:688)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:669)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.ChannelOutboundHandlerAdapter.flush(ChannelOutboundHandlerAdapter.java:115)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:688)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:718)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:706)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:741)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.DefaultChannelPipeline.writeAndFlush(DefaultChannelPipeline.java:895)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.AbstractChannel.writeAndFlush(AbstractChannel.java:240)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at net.minecraft.server.v1_8_R3.NetworkManager$4.run(NetworkManager.java:198)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:380)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:268)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116)
[18:37:16] [Netty Epoll Server IO #3/WARN]: 	at java.lang.Thread.run(Thread.java:745)
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Server thread/INFO]: player lost connection: Internal Exception: java.io.IOException: Error while write(...): Broken pipe
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
[18:37:16] [Netty Epoll Server IO #3/WARN]: java.nio.channels.ClosedChannelException
```

and it's a setup with Spigot behind latest BungeeCord without protocollib.

**Comment by gigosaurus** [comment]
Here's the stacktrace I get from BungeeCord's console when this happens for me:

    io.netty.handler.codec.DecoderException: java.lang.IndexOutOfBoundsException: readerIndex(0) + length(1) exceeds writerIndex(0): SlicedAbstractByteBuf(ridx: 0, widx: 0, cap: 0/0, unwrapped: PooledUnsafeDirectByteBuf(ridx: 2, widx: 8, cap: 336))
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:99)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:318)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:304)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:318)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:304)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:276)
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:354)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:244)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:318)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:304)
        at io.netty.handler.timeout.ReadTimeoutHandler.channelRead(ReadTimeoutHandler.java:152)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:318)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:304)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:846)
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:823)
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:339)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:255)
        at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:112)
        at java.lang.Thread.run(Thread.java:745)
    Caused by: java.lang.IndexOutOfBoundsException: readerIndex(0) + length(1) exceeds writerIndex(0): SlicedAbstractByteBuf(ridx: 0, widx: 0, cap: 0/0, unwrapped: PooledUnsafeDirectByteBuf(ridx: 2, widx: 8, cap: 336))
        at io.netty.buffer.AbstractByteBuf.checkReadableBytes0(AbstractByteBuf.java:1178)
        at io.netty.buffer.AbstractByteBuf.readByte(AbstractByteBuf.java:587)
        at net.md_5.bungee.protocol.DefinedPacket.readVarInt(DefinedPacket.java:81)
        at net.md_5.bungee.protocol.DefinedPacket.readVarInt(DefinedPacket.java:71)
        at net.md_5.bungee.protocol.MinecraftDecoder.decode(MinecraftDecoder.java:28)
        at net.md_5.bungee.protocol.MinecraftDecoder.decode(MinecraftDecoder.java:10)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:89)
        ... 19 more

**Comment by Paulomart** [comment]
it musst be a change in 318db73f3c7b48eaa5421cd4da460950fc5388a7.

**Comment by MylesIsCool** [comment]
It's worth noting the internals of netty are so vast it's very hard when converting from one system to another to ensure same results. I'm about to push the patch and it's down to one of these issues.

Fixed in 0.4.2
